File: 09/main.py
from __future__ import annotations
from typing import Dict, List, NamedTuple, FrozenSet, Set, Iterator


class Point(NamedTuple):
    row: int
    col: int

    def neighbors(self):
        yield Point(self.row, self.col + 1)
        yield Point(self.row - 1, self.col)
        yield Point(self.row, self.col - 1)
        yield Point(self.row + 1, self.col)


class HeightMap:
    def __init__(self, heightmap):
        self.heightmap = heightmap

    def __getitem__(self, key):
        return self.heightmap[key]

    @property
    def width(self):
        return len(self.heightmap[0])

    @property
    def height(self):
        return len(self.heightmap)

    def neighbors(self, point: Point) -> Iterator[Point]:
        for n in point.neighbors():
            if (0 <= n.row < self.height) and (0 <= n.col < self.width):
                yield n

    def find_low_points(self):
        low_points = []
        for ri, row in enumerate(self.heightmap):
            for ci, height in enumerate(row):
                point = Point(ri, ci)
                if all(
                    height < self.heightmap[n.row][n.col] for n in self.neighbors(point)
                ):
                    low_points.append(point)
        return low_points

    def get_risk_level(self, point: Point):
        return self.heightmap[point.row][point.col] + 1


def read_puzzle_input(filename: str) -> HeightMap:
    with open(filename) as stream:
        return HeightMap(
            [[int(n) for n in row] for row in stream.read().strip().split("\n")]
        )


if __name__ == "__main__":
    puzzle_input: HeightMap = read_puzzle_input("input.txt")
    print(sum(puzzle_input.get_risk_level(p) for p in puzzle_input.find_low_points()))

File: 09/main.hs
import Data.Maybe (catMaybes, isNothing)
import Data.Map (Map)
import qualified Data.Map as M
import Data.Set (Set)
import qualified Data.Set as S
import Data.List (sort)

data Point = Point Int Int deriving (Show, Eq)

instance Ord Point where
    (Point x1 y1) `compare` (Point x2 y2)
        | y1 == y2 = x1 `compare` x2
        | otherwise = y1 `compare` y2

enumerate = zip [0..]

melt :: [[Int]] -> Map Point Int
melt xs = M.fromList $ concat $ [[(Point i j, height) | (j, height) <- enumerate row] | (i, row) <- enumerate xs]

parseInput :: String -> Map Point Int
parseInput string = melt $ map readLine (lines string)
  where
    readLine = map (read . (: ""))

getNeighbors :: Point -> [Point]
getNeighbors (Point row col) = [Point (row - 1) col, Point (row + 1) col, Point row (col - 1), Point row (col + 1)]

getNeighborHeights :: Map Point Int -> Point -> [Int]
getNeighborHeights heightMap point = catMaybes $ [M.lookup neighbor heightMap | neighbor <- getNeighbors point]

getLowPoints :: Map Point Int -> [Point]
getLowPoints heightMap = [point | (point, height) <- M.toList heightMap, all (>height) (getNeighborHeights heightMap point)]

getRiskLevel :: Map Point Int -> Int
getRiskLevel heightMap = sum $ map (+1) lowHeights
    where lowHeights = catMaybes [M.lookup lowPoint heightMap | lowPoint <- getLowPoints heightMap]

_getBasinAroundLowPoint :: Map Point Int -> Set Point -> [Point] -> Set Point -> Set Point
_getBasinAroundLowPoint heightMap basin [] checked = basin
_getBasinAroundLowPoint heightMap basin (p:rest) checked
    | isNothing height = _getBasinAroundLowPoint heightMap basin rest newChecked
    | height == Just 9 = _getBasinAroundLowPoint heightMap basin rest newChecked
    | otherwise = _getBasinAroundLowPoint heightMap newBasin (rest ++ neighbors) newChecked
    where newChecked = S.insert p checked
          height = M.lookup p heightMap
          newBasin = S.insert p basin
          neighbors = filter (`S.notMember` checked) . filter (`M.member` heightMap) $ getNeighbors p

getBasinAroundLowPoint :: Map Point Int -> Point -> Set Point
getBasinAroundLowPoint heightMap lowPoint = _getBasinAroundLowPoint heightMap basin neighbors checked
    where lowPointSet = S.singleton lowPoint
          neighbors = getNeighbors lowPoint
          basin = lowPointSet
          checked = lowPointSet

getBasins :: Map Point Int -> [Set Point]
getBasins heightMap = map (getBasinAroundLowPoint heightMap) (getLowPoints heightMap)

getProductOf3LargestBasins :: Map Point Int -> Int
getProductOf3LargestBasins = product . take 3 . reverse . sort . map length . getBasins

main :: IO ()
main = do
    content <- readFile "input.txt"
    let heightMap = parseInput content
    print (getRiskLevel heightMap)
    print (getProductOf3LargestBasins heightMap)
